### ISSUE TITLE
[Bug]: The output_neurons dimension isn't consistent with SentencePiece vocabulary size when training a model using a character based tokenizer 

### ISSUE THREAD

**Issue by qanastek** [speechbrain/speechbrain#1696]
### Describe the bug

Dear maintainers,

My coworkers and I were confronted multiple time to a bug caused by an inconsistency between the dimension of the `output_neurons` and the vocabulary size of the SentencePiece character based tokenizer.

Let me explains by taking for example the recipe for `Wav2Vec + CTC` for CommonVoice.

If we are training an ASR model by setting up the tokenizer to produce a maximum of 100 characters by changing the parameter `model_type` to `char` and `vocab_size` to `100` in the SpeechBrain implementation of SentencePiece.

We will obtain an `NaN` loss after a few epochs if the real number of characters present in the training subset is higher than the `output_neurons` since the implementation of the SentencePiece present in SpeechBrain exclude the parameter `--vocab_size=<output_neurons>` in the case of the characters based model at the line 306. 

https://github.com/speechbrain/speechbrain/blob/11c70de96b3cc33ccd31fad7e470297e5c0ca69f/speechbrain/tokenizers/SentencePiece.py#L279-L314

If I change in this recipe, the `output_neurons` to very low number like `15` and the `model_type` to `char`. I will technically train a model with only 15 outputs neurons, but my tokenizer will encode and give me as inputs the documents with `68` characters. Which will progressively conduct to a `NaN` loss since the model try to produce unknown characters.

### Expected behaviour

I expect the character based tokenizer to limit the number of characters to the given `vocab_size`.

### To Reproduce

1. Take the Wav2Vec + CTC recipe for CommonVoice [here](https://github.com/speechbrain/speechbrain/blob/develop/recipes/CommonVoice/ASR/CTC/hparams/train_fr_with_wav2vec.yaml)
2. Just replace `output_neurons` by a small number like `15` and `token_type` to `char`.
3. Enjoy the `NaN` loss after 7 to 10 epochs.

### Versions

Up to date with the `default` branch.

### Relevant log output

```shell
Image below in "Additional context".
```

### Additional context

![image (6)](https://user-images.githubusercontent.com/19389475/201071743-2f7eb523-cb36-4c45-aef3-1ae02bfd9196.png)

**Comment by TParcollet** [comment]
Hi, I believe that the issue comes from sentencepiece. When using char as a unit, it will automatically create a vocabulary containing each individual char without any respect for the given output size, which is expected. This means that from CommonVoice version to another, things will break because the vocabulary will be different. Unfortunately, with most recipes, the PyTorch model is built BEFORE running the sentencepiece, which is also expected as all initialisation should be done before running anything. This behavior could maybe be change ... but I am not sure how. In SpeechBrain, we don't like having parameters that are magically or dynamically discovered and used. We largely prefer people first training a char sentencepiece, then look at the .vocab file, and adjust the output_neurons accordingly. This forces the user to be aware of what is happening in the pipeline.

**Comment by AsuMagic** [comment]
Likely duplicate of #1589

**Comment by Adel-Moumen** [comment]
Fixed in #1852
